### ISSUE TITLE
--test fails with `error: no test target named`

### ISSUE THREAD

**Issue by ede1998** [flamegraph-rs/flamegraph#139]
I'm trying to run `cargo flamegraph` for the unit tests in my crate. Unfortunately, it always fails with this error:
`error: no test target named {value}`
I tried numerous variations of the `--test` flag in combination with the binary name, path, test name, test full names, ... but nothing seems to work.

Is the `--test` feature not working right now or am I using it wrong? If I am using it wrong, can we maybe get a usage example in the README?

Workaround:
I manually build the test executable with `cargo build --tests --release` and then ran `flamegraph "target/release/deps/test-executable-9fb133aef8031813 -- full::test::name`

**Comment by djc** [comment]
I think it does work to some extent, or there would likely be more issues filed. Are you using the latest version? Would you mind digging in a little to figure out why it's failing? You could try to pass `-v` to get verbose output that prints the build command.

**Comment by ede1998** [comment]
I think I found the reason:
`cargo flamegraph --test <name>` invokes `cargo build --release --test <name>`, which fails if you invoke it with a unit test name. [The Cargo Book](https://doc.rust-lang.org/cargo/commands/cargo-build.html) says that `--test <name>` is used to "build the specified integration test". It seems to me like `cargo flamegraph` cannot run unit tests right now and the workaround I described above is needed.
After looking a bit into the `cargo build` documentation, I didn't find a way to use `cargo build` to just build the unit test executable. Is this even possible? `--tests` builds all tests, integration tests and unit tests alike.

Answering your questions:

> Are you using the latest version?

I think so, just installed it yesterday using cargo. `cargo flamegraph --version` prints `0.4.0`.

> Would you mind digging in a little to figure out why it's failing?

I created a test project to reproduce it: `cargo new flamegraph-test`. I added a unit test to `main.rs`:
```rust
use std::time::Duration;

fn main() {
    println!("Hello, world!");
}

#[test]
fn test_run() {
    std::thread::sleep(Duration::from_secs(3));
}
```
When I run `cargo flamegraph --test test_run`, I get the error `error: no test target named 'test_run'`.

> You could try to pass -v to get verbose output that prints the build command.

After adding the `-v` flag, flamegraph prints `build command: "cargo" "build" "--release" "--test" "test_run"`. Trying to run this command directly yields the same `error: no test target named 'test_run'` error.

**Comment by djc** [comment]
Ah yeah, this is expected behavior. The way unit tests work is that they're all stuffed into a single binary together, and cargo flamegraph currently doesn't have a way to execute the unit tests. If you want to contribute a PR to change that (probably including the infrastructure needed to pass a test selection on to the executed workload command), that would be great!

**Comment by ede1998** [comment]
I started working on the PR today. However I have some questions:

Why is `--message-format=json` used for checking if building with debuginfo but not for finding the executable path?
The json that cargo produces seems to contain the path to the correct executables directly. The current approach of hand-crafting the executable path seems to lead to issues: https://github.com/flamegraph-rs/flamegraph/issues/79
Is this just something that has historically grown or is there a deeper reason behind it?

The second question is about the design of the PR: I realized that a simple switch `--unit-test test_name` is insufficient. A single package can have multiple targets (lib, bin1, bin2, ...) and each of those targets has its own unit test binary. If the switch just consumes a test name, we don't know the test binary to execute. Therefore I was thinking to just pass the target name after the switch and the test name as additional trailing argument like so: `cargo flamegraph --unit-test binary1 -- test::in:binary::one`. This would also simplify the handling of the test selection as we would just have the test executable handle it by passing it unchanged in `trailing_arguments`. However, it might confused first time users that the test has to be specified after `--` and not as argument to `--unit-test`.
(If the package only has a single target, the call could be simplified to just `cargo flamegraph --unit-test -- test::in::only::target`.)
What do you think about this design? Any suggestions/alternatives?

Third question: I didn't find a way yet to build only the test executable. I used `--tests` as a fallback but it builds all tests, including integration tests. Do you happen to know how/if I can tell cargo to only build a specific unit test executable?

**Comment by ede1998** [comment]
I implemented a very basic version that I will have to refine:
https://github.com/ede1998/flamegraph/commit/b1da2d012d04de3403638254dd55c2612c83210d

I used the design I proposed where the argument to `--unit-test` is the target and not the test name.

Also, I think I found a solution for my tthird question: `cargo test --no-run --lib` or `cargo test --no-run --bin <binname>` builds the unit test binary. This should also avoid issues if the user has `test = false` set for the target in `Cargo.toml`. With `--tests` this wouldn't work.

**Comment by djc** [comment]
As for your first question: I think that's probably just a historical artefact. Getting the executable path from the JSON seems likely to be superior (and a chance to delete a bunch of finicky code). Would be happy to have a PR for this, too, ideally a separate one?

Your proposed design seems okay to me. Would `cargo flamegraph --unit-test binary1 test::in::binary::one` (that is, without the explicit `--`) work too? If not, would be nice to be explicit about that in any user facing documentation/examples.

**Comment by ede1998** [comment]
> Would `cargo flamegraph --unit-test binary1 test::in::binary::one` (that is, without the explicit `--`) work too?

It never occurred to me to even try. I just thought that the `--` is needed to separate the `cargo flamegraph` arguments from the executable arguments. After trying it seems that it doesn't work because `clap` tries to group the test as a 2nd value to `--unit-test` instead of using it as a positional argument. If another flag separates the positional argument from the `--unit-test` flag, `--` can be omitted:
```
$ cargo flamegraph --unit-test binary1 test::in::binary::one
error: The value 'test::in::binary::one' was provided to '--unit-test <unit-test>', but it wasn't expecting any more values

$ cargo flamegraph --unit-test mainnotest --verbose test::in::binary::one
build command: "cargo" "build" "--release" "--tests"
    Finished release [optimized + debuginfo] target(s) in 0.04s
```

> Would be happy to have a PR for this

I'll start working on that as well because I noticed another issue with the running flamegraph for unit-tests. I can't tell which executable is the unit-test executable and which is the normal executable, as both files only differ in the random suffix string.
```
$ cargo build --release; cargo build --release --tests
$ ls target/release/deps
main1-7d658e30325341e3 main1-7d658e30325341e3
```
For now, I'll probably just take the executable with the newest timestamp. Alternatively, we could run `cargo clean` before building. But to make it foolproof, it's neccessary to read the output of `cargo build` in my opinion.

**Comment by ede1998** [comment]
implemented in #140
